### ISSUE TITLE
Addition of Siamese loss or contrastive loss function

### ISSUE THREAD

**Issue by Hemantr05** [pytorch/pytorch#34879]
## 🚀 Feature

I would like to contribute to the pytorch by adding the [ contrastive loss](http://yann.lecun.com/exdb/publis/pdf/hadsell-chopra-lecun-06.pdf) function which is usually implemented  for siamese networks

## Motivation

I have been working on siamese networks and have implemented constrastive loss to measure similarity between bottleneck embeddings of the autoencoder.

## Pitch

The output of the loss function is similar to that of  [this](https://github.com/pytorch/pytorch/blob/d9b97a4ffd3d65eeaccb5f8729926976fa868981/torch/nn/functional.py#L3544) but for siamese network instead of triplet network.

It returns a float value signifying the distance between the bottleneck embedding.

## Additional context

This loss implemented for all siamese network implementations (such as facial similarity, Omniglot examples)

**Comment by Hemantr05** [comment]
@mrshenli  can create a PR?

**Comment by Hemantr05** [comment]
May I add the loss function?

**Comment by mrshenli** [comment]
Adding this to triage review meeting to discuss how can we should proceed with this feature request and contribution.

**Comment by ezyang** [comment]
This one should go to the "needs research" queue

**Comment by gchanan** [comment]
1430 citations since 2006.

**Comment by Hemantr05** [comment]
> 1430 citations since 2006.

Yes

**Comment by Hemantr05** [comment]
The loss is very widely used for siamese networks

**Comment by Hemantr05** [comment]
@pbelevich @ezyang @gchanan @mrshenli What's the conclusion sir/madam?

**Comment by soumith** [comment]
it seems like a reasonable addition.

The loss ends up being something like [reference](https://github.com/harveyslash/Facial-Similarity-with-Siamese-Networks-in-Pytorch/blob/master/Siamese-networks-medium.ipynb):

```
euclidean_distance = F.pairwise_distance(output1, output2, keepdim = True)
loss_contrastive = torch.mean((1-label) * torch.pow(euclidean_distance, 2) +
                                      (label) * torch.pow(torch.clamp(self.margin - euclidean_distance, min=0.0), 2))
```

Let's use the API `F.siamese_loss` and `nn.SiameseLoss` unless you have better ideas. I think ContrastiveLoss is too broad of a term, as contrastive losses are of various kinds.

As with all contributions, we would need unit tests, docstrings etc.

thanks for your effort.

**Comment by Hemantr05** [comment]
> it seems like a reasonable addition.
> 
> The loss ends up being something like [reference](https://github.com/harveyslash/Facial-Similarity-with-Siamese-Networks-in-Pytorch/blob/master/Siamese-networks-medium.ipynb):
> 
> ```
> euclidean_distance = F.pairwise_distance(output1, output2, keepdim = True)
> loss_contrastive = torch.mean((1-label) * torch.pow(euclidean_distance, 2) +
>                                       (label) * torch.pow(torch.clamp(self.margin - euclidean_distance, min=0.0), 2))
> ```
> 
> Let's use the API `F.siamese_loss` and `nn.SiameseLoss` unless you have better ideas. I think ContrastiveLoss is too broad of a term, as contrastive losses are of various kinds.
> 
> As with all contributions, we would need unit tests, docstrings etc.
> 
> thanks for your effort.

F.siamese_loss and nn.SiameseLoss is perfect.
Thanks for considering it

Will perform and provide unit tests, docstrings, etc..

Thank you for your support

**Comment by henrywallace** [comment]
See also `ContrastiveLoss` implementation from https://github.com/KevinMusgrave/pytorch-metric-learning/blob/aa0e4dd375d9c233eddd184835ff2a51b2e59102, among others.

**Comment by Hemantr05** [comment]
> See also `ContrastiveLoss` implementation from https://github.com/KevinMusgrave/pytorch-metric-learning/blob/aa0e4dd375d9c233eddd184835ff2a51b2e59102, among others.

Will do @henrywallace
